x/mongo/driver/topology/connection.go

```diff
@@ -164,35 +193,17 @@ func (c *connection) connect(ctx context.Context) (err error) {
 	// cancellation still applies but with an added timeout to ensure the connectTimeoutMS option is applied to socket
 	// establishment and the TLS handshake as a whole. This is created outside of the connectContextMutex lock to avoid
 	// holding the lock longer than necessary.
-	c.connectContextMutex.Lock()
-	var handshakeCtx context.Context
-	handshakeCtx, c.cancelConnectContext = context.WithCancel(ctx)
-	c.connectContextMutex.Unlock()
+	ctx, cancel := contextWithNewConnTimeout(ctx, c)
+	defer cancel()
 
-	dialCtx := handshakeCtx
-	var dialCancel context.CancelFunc
-	if c.config.connectTimeout != 0 {
-		dialCtx, dialCancel = context.WithTimeout(handshakeCtx, c.config.connectTimeout)
-		defer dialCancel()
-	}
+	go func() {
+		defer cancel()
 
-	defer func() {
-		var cancelFn context.CancelFunc
-
-		c.connectContextMutex.Lock()
-		cancelFn = c.cancelConnectContext
-		c.cancelConnectContext = nil
-		c.connectContextMutex.Unlock()
-
-		if cancelFn != nil {
-			cancelFn()
-		}
+		<-c.cancelConnSig
```

Comment:
`cancelConnSig` is never initialized, so this receive will block forever. Also, if `closeConnectContext` is never called, this will leak a goroutine. I suggest two changes:
* Initialize `cancelConnSig` with `make(chan struct{}, 1)` in `newConnection`.
  * Use a buffered channel to resolve race conditions between starting this goroutine and calling `closeConnectContext`.
* Add `defer conn.closeConnectContext` to `conn.connect` so we don't leak this goroutine.

Consider moving the following logic into the beginning of the `connection.close` method to reduce the chances for bugs when closing a connection:
```go
func (c *connection) close() error {
	conn.closeConnectContext()
	conn.wait()

	// ...
}
```
In the case the connection is already connected, those calls are idempotent and shouldn't block. That would allow us to make all usage of the `cancelConnSig` internal to the `connection` type.

Reply:
Good catch! This pattern seems fine, except we won't be able to wait for the connection to be done before closing the connection. Otherwise close() will block indefinitely whenever there is a timeout while making a handshake with the server.